### DISCUSSION TITLE
Window rules questions

### DISCUSSION THREAD

**Discussion by stefonarch** [labwc/labwc#924] [Q&A] [answered]
Coming  back I've seen window rules enabled. Not sure if I miss something but 
```
  <windowRules>
      <windowRule identifier="lxqt-panel">
        🔗<action name="MoveToEdge" direction="up" />
      </windowRule
   </windowRules>
```
does nothing, tried also with "normal" applications. 

Second: I tried blindly
 ```
 <windowRules>

      <!-- Property -->
      <windowRule identifier="lxqt-*" alwaysonTop="yes"/>

    </windowRules>
```
which doesn't work. I didn't find a list with available properties anywhere.
Thanks in advance.

**Comment by Consolatis** [comment, marked as answer]
In your first example there is a `🔗` anchor character before the opening action tag. That is something I didn't think about when we added them to the labwc.github.io man pages. I have an idea on how to fix that so they don't show up when copying something. You can test with the `Maximize` action to see if the rule matches in the first place. It is possible that some actions don't work properly as they are called directly after mapping the window, this requires some further testing.

For the 2nd example: so far labwc only supports the `serverDecoration` property for window rules.
See https://labwc.github.io/labwc-config.5.html#window_rules

**Reply by stefonarch** [reply]
Thanks, I just didn't notice, in featherpad there is a little square box. 
This said the rule works perfectly for lxqt-panel (which still would need at least "alwaysOnTop" and "onAllWorkspaces )

**Reply by Consolatis** [reply]
You should be able to just call the `ToggleAlwaysOnTop` action within the rule (which currently also means "onAllWorkspaces"). You can have multiple actions within each rule.

**Comment by stefonarch** [comment]
Not sure if that is a bug, I tried "down" and it's not quite precise leaving a gap:

![20230516_14h38m11s_grim](https://github.com/labwc/labwc/assets/10681413/69029871-7413-4f37-befa-a18b4853f49b)

"left" is on the border, "right" has a bigger gap than "down".

**Reply by Consolatis** [reply]
Do you have `<core><gap>` set by chance? That wouldn't explain the different gaps of "left" and "right" though.
You could try to bind `MoveToEdge` to some keybind and call it manually when the panel has focus to make sure it has nothing to do with the rule calling it directly after the map.

**Comment by stefonarch** [comment]
gap is 0, with the keybind I see the same result with the panel and also  `lxqt-runner` but not with application windows.

**Comment by johanmalm** [comment]
@stefonarch I'm playing with window rules for `lxqt-panel` and `pcmanfm-qt --desktop`. I can get them both working quite well, but haven't decided on how to best implement rules for them. At the moment it's just a complete non-pushable hack, so can't share yet.

For example, I've added a `SendToLayer` action to move them to output specific layer-shell layers. Seems to work okay (except for popups which need to be reparented so that they render above normal windows).

I started off by adding an `AlwaysOnBottom` layer (for all outputs), but the problem with that approach is that the layer-shell layers wrap around, so I could never get `pcmanfm-qt --desktop` below the layer-shell bottom/background layers which is a bit annoying if you have a panel in the bottom-layer.

Have also added `<margin>` to manually set usable-area.

In my current hack, I've just used `SnapToRegion` to position the panel (the only problem being that I can only specify the height as a percentage, but I'm not too worried about that for the minute). We could easily add a `MoveTo` action or something, but I just want to think through the bigger approach first.

I'm distracted by real-life for week or so, but will pushing something when it's ready for play + critique.

**Reply by stefonarch** [reply]
That's nice to hear! In fact I was about to ask if there is any attempt in those directions, as the wayfire [ lxqt-desktop-shell](https://gitlab.com/wayfireplugins/lxqt-desktop-shell) plugin does work almost perfectly, and in labwc the same could probably quite  easily achieved which would provide at least a desktop with icon and so on. `SendToLayer` is infact the thing needed, together maybe with `SetRole`.

The layer for the panel is` layer_top` here, this works perfectly for fullscreen (gets covered), for the exclusive area there is a code in the plugin.

> I'm distracted by real-life for week or so, but will pushing something when it's ready for play + critique.

Can't wait ;)

**Comment by stefonarch** [comment]
@Consolatis I found the reason for the gap with "MoveToEdge" right|down: there is apparently a 60px limit in the code for windows and it counts from the upper edge I guess: the gap appears with lxqt-panel's height under 60px only. 60px is quite huge for it.

**Reply by Consolatis** [reply]
Right, we wanted to switch that hardcoded limit to the SSD height + border height (if enabled, `1` otherwise) for a while but didn't follow through so far. For other panels its not really an issue because this restriction only applies to usual windows but not to layershell surfaces.

**Reply by johanmalm** [reply]
Will have a look at `lxqt-panel` soon. Yes, makes sense to remove that hard-coded value like Consolatis said.
